homeassistant/components/portainer/switch.py

```diff
@@ -0,0 +1,124 @@
+
+from homeassistant.components.switch import (
+    SwitchDeviceClass,
+    SwitchEntity,
+    SwitchEntityDescription,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+
+from . import PortainerConfigEntry
+from .coordinator import PortainerCoordinator
+from .entity import PortainerContainerEntity, PortainerCoordinatorData
+
+
+@dataclass(frozen=True, kw_only=True)
+class PortainerSwitchEntityDescription(SwitchEntityDescription):
+    """Class to hold Portainer switch description."""
+
+    is_on_fn: Callable[[DockerContainer], bool | None]
+    turn_on_fn: Callable[[Portainer, int, str], Coroutine[Any, Any, None]]
+    turn_off_fn: Callable[[Portainer, int, str], Coroutine[Any, Any, None]]
+
+
+async def stop_container(
+    portainer: Portainer, endpoint_id: int, container_id: str
+) -> None:
+    """Stop a container."""
+    await portainer.stop_container(endpoint_id, container_id)
+
+
+async def start_container(
+    portainer: Portainer, endpoint_id: int, container_id: str
+) -> None:
+    """Start a container."""
+    await portainer.start_container(endpoint_id, container_id)
+
+
+SWITCHES: tuple[PortainerSwitchEntityDescription, ...] = (
+    PortainerSwitchEntityDescription(
+        key="container",
+        translation_key="container",
+        device_class=SwitchDeviceClass.SWITCH,
+        is_on_fn=lambda data: data.state == "running",
+        turn_on_fn=start_container,
+        turn_off_fn=stop_container,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: PortainerConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up Portainer switch sensors."""
+
+    coordinator: PortainerCoordinator = entry.runtime_data
+
+    async_add_entities(
+        PortainerContainerSwitch(
+            coordinator=coordinator,
+            entity_description=entity_description,
+            device_info=container,
+            via_device=endpoint,
+        )
+        for endpoint in coordinator.data.values()
+        for container in endpoint.containers.values()
+        for entity_description in SWITCHES
+    )
+
+
+class PortainerContainerSwitch(PortainerContainerEntity, SwitchEntity):
+    """Representation of a Portainer container switch."""
+
+    entity_description: PortainerSwitchEntityDescription
+
+    def __init__(
+        self,
+        coordinator: PortainerCoordinator,
+        entity_description: PortainerSwitchEntityDescription,
+        device_info: DockerContainer,
+        via_device: PortainerCoordinatorData,
+    ) -> None:
+        """Initialize the Portainer container switch."""
+        self.entity_description = entity_description
+        super().__init__(device_info, coordinator, via_device)
+
+        device_identifier = (
+            self._device_info.names[0].replace("/", " ").strip()
+            if self._device_info.names
```

Comment:
Should fix this typing issue in a follow-up so the if-statement can be dropped. I take it. it can never be empty.